File: compatibility/mysql/node/mysql_test.js
const mysql = require('mysql');
const fs = require('fs');

class Test {
    constructor(query, expectedResults) {
        this.query = query;
        this.expectedResults = expectedResults;
    }

    async run(connection) {
        return new Promise((resolve, reject) => {
            console.log(`Running test: ${this.query}`);
            connection.query(this.query, (error, results, fields) => {
                if (error) {
                    console.error(error.message);
                    return resolve(false);
                }
                if (results.length === undefined || results.length === 0) {
                    if (this.expectedResults.length !== 0) {
                        console.error(`Expected ${this.expectedResults.length} rows, got 0`);
                        return resolve(false);
                    }
                    console.log("Returns 0 rows");
                    return resolve(true);
                }
                if (fields.length !== this.expectedResults[0].length) {
                    console.error(`Expected ${this.expectedResults[0].length} columns, got ${fields.length}`);
                    return resolve(false);
                }
                for (let i = 0; i < results.length; i++) {
                    const row = results[i];
                    const expectedRow = this.expectedResults[i];
                    for (let j = 0; j < expectedRow.length; j++) {
                        const columnName = fields[j].name;
                        if (String(row[columnName]) !== String(expectedRow[j])) {
                            console.error(`Expected: ${String(expectedRow[j])}, got: ${String(row[columnName])}`);
                            return resolve(false);
                        }
                    }
                }
                console.log(`Returns ${results.length} rows`);
                if (results.length !== this.expectedResults.length) {
                    console.error(`Expected ${this.expectedResults.length} rows`);
                    return resolve(false);
                }
                resolve(true);
            });
        });
    }
}

class Tests {
    constructor() {
        this.connection = null;
        this.tests = [];
    }

    connect(ip, port, user, password) {
        this.connection = mysql.createConnection({
            host: ip,
            port: port,
            user: user,
            password: password
        });
        return new Promise((resolve, reject) => {
            this.connection.connect((err) => {
                if (err) {
                    console.error('Error connecting to the database:', err.message);
                    return reject(err);
                }
                resolve();
            });
        });
    }

    disconnect() {
        return new Promise((resolve, reject) => {
            this.connection.end((err) => {
                if (err) {
                    console.error('Error disconnecting from the database:', err.message);
                    return reject(err);
                }
                resolve();
            });
        });
    }

    addTest(query, expectedResults) {
        this.tests.push(new Test(query, expectedResults));
    }

    async runTests() {
        for (const test of this.tests) {
            if (!(await test.run(this.connection))) {
                return false;
            }
        }
        return true;
    }

    readTestsFromFile(filename) {
        const data = fs.readFileSync(filename, 'utf8');
        const lines = data.split('\n');
        let query = null;
        let expectedResults = [];
        for (const line of lines) {
            const trimmedLine = line.trim();
            if (!trimmedLine) {
                if (query) {
                    this.addTest(query, expectedResults);
                    query = null;
                    expectedResults = [];
                }
            } else if (query === null) {
                query = trimmedLine;
            } else {
                expectedResults.push(trimmedLine.split(','));
            }
        }
        if (query) {
            this.addTest(query, expectedResults);
        }
    }
}

(async () => {
    if (process.argv.length < 6) {
        console.error("Usage: node mysql_test.js <ip> <port> <user> <password> <testFile>");
        process.exit(1);
    }

    const tests = new Tests();
    await tests.connect(process.argv[2], parseInt(process.argv[3]), process.argv[4], process.argv[5]);
    tests.readTestsFromFile(process.argv[6]);

    if (!(await tests.runTests())) {
        await tests.disconnect();
        process.exit(1);
    }
    await tests.disconnect();
})();

File: compatibility/mysql/python/mysql_test.py
import mysql.connector

class MySQLTest:
    class Test:
        def __init__(self, query, expected_results):
            self.query = query
            self.expected_results = expected_results

        def run(self, cursor):
            print(f"Running test: {self.query}")
            cursor.execute(self.query)
            if cursor.description is None:
                if len(self.expected_results) != 0:
                    print(f"Expected {len(self.expected_results)} rows, got 0")
                    return False
                print("Returns 0 rows")
                return True

            rows = cursor.fetchall()
            if len(rows[0]) != len(self.expected_results[0]):
                print(f"Expected {len(self.expected_results[0])} columns, got {len(rows[0])}")
                return False

            for row, expected_row in zip(rows, self.expected_results):
                if list(map(str, row)) != list(map(str, expected_row)):
                    print(f"Expected: {list(map(str, expected_row))}, got: {list(map(str, row))}")
                    return False

            print(f"Returns {len(rows)} rows")
            if len(rows) != len(self.expected_results):
                print(f"Expected {len(self.expected_results)} rows")
                return False

            return True

    def __init__(self):
        self.conn = None
        self.tests = []

    def connect(self, ip, port, user, password):
        try:
            self.conn = mysql.connector.connect(
                host=ip,
                port=port,
                user=user,
                password=password
            )
        except Exception as e:
            raise RuntimeError(e)

    def disconnect(self):
        self.conn.close()

    def add_test(self, query, expected_results):
        self.tests.append(self.Test(query, expected_results))

    def run_tests(self):
        for test in self.tests:
            cursor = None
            try:
                cursor = self.conn.cursor()
                if not test.run(cursor):
                    return False
            except Exception as e:
                print(f"Error running test: {e}")
                return False
            finally:
                if cursor:
                    cursor.close()
        return True

    def read_tests_from_file(self, filename):
        with open(filename, "r") as file:
            lines = file.readlines()

        query = None
        expected_results = []
        for line in lines:
            line = line.strip()
            if not line:
                if query:
                    self.add_test(query, expected_results)
                    query = None
                    expected_results = []
            elif query is None:
                query = line
            else:
                expected_results.append(line.split(","))

        if query:
            self.add_test(query, expected_results)

if __name__ == "__main__":
    import sys
    if len(sys.argv) != 6:
        print(f"Usage: {sys.argv[0]} <ip> <port> <user> <password> <testFile>")
        sys.exit(1)

    mysql_test = MySQLTest()
    mysql_test.connect(sys.argv[1], int(sys.argv[2]), sys.argv[3], sys.argv[4])
    mysql_test.read_tests_from_file(sys.argv[5])

    if not mysql_test.run_tests():
        mysql_test.disconnect()
        sys.exit(1)

    mysql_test.disconnect()